[CAS] Adopt `UnifiedOnDiskCache` for the on-disk CAS for clang

akyrtzi opened the pull request:
Notable changes:

* Having a separate flag for on-disk path for `ActionCache` becomes redundant (only one on-disk path for CAS/cache configuration is sufficient), so `-faction-cache-path` and related flags of testing tools are removed.
* To accomodate testing, other testing tools adopt `UnifiedOnDiskCache` as well.
* `llvm-cas` gets new functionality to accomodate testing:
  * `-import`: import objects from another CAS
  * `-put-cache-key`: set a value for a cache key

Note that clang opens a `UnifiedOnDiskCache` without imposing any size limit, it is assumed that a higher level service, e.g. the build system or the clang daemon, is going to be responsible for managing growth of the CAS databases during a build.

akyrtzi committed:
[CAS] Adopt `UnifiedOnDiskCache` for the on-disk CAS for clang

Notable changes:

* Having a separate flag for on-disk path for `ActionCache` becomes redundant (only one on-disk path for CAS/cache
  configuration is sufficient), so `-faction-cache-path` and related flags of testing tools are removed.
* To accomodate testing, other testing tools adopt `UnifiedOnDiskCache` as well.
* `llvm-cas` gets new functionality to accomodate testing:
  * `-import`: import objects from another CAS
  * `-put-cache-key`: set a value for a cache key

Note that clang opens a `UnifiedOnDiskCache` without imposing any size limit, it is assumed that a higher level service,
e.g. the build system or the clang daemon, is going to be responsible for managing growth of the CAS databases during a build.

akyrtzi commented:
@swift-ci Please test

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan commented on clang/lib/CAS/CASOptions.cpp:
```diff
@@ -73,20 +72,17 @@ void CASOptions::freezeConfig(DiagnosticsEngine &Diags) {
-  std::string Path = Config.CachePath;
-  if (Path == "auto")
-    Path = getDefaultOnDiskActionCachePath();
+  std::string Path = getDefaultOnDiskActionCachePath();
```
I will ditch this for something new like `getClangDefaultCachePath()`. No need to expose the creation for ObjectStore and ActionCache here.

akyrtzi replied:
Are you suggesting introducing a `getClangDefaultCachePath()` function which then calls `getDefaultOnDiskActionCachePath()`, instead of calling the latter directly?

cachemeifyoucan replied:
I think since we decide to bundle object store and action cache together in clang, we should just avoid create them separately. All we need is pick a default path and initialize them together.

Also what do you think if we just get rid of `auto` and InMemory implementation and use empty string as default location?

akyrtzi replied:
Yeah, that makes sense. We can also replace `CASOptions::getOrCreateObjectStore` and `CASOptions::getOrCreateActionCache` with one call to return both.

akyrtzi replied:
> Also what do you think if we just get rid of auto and InMemory implementation and use empty string as default location?

After looking into this, it's not straightforward because of 2 reasons:
* There's code like `Options.MCOptions.CAS = CASOpts.getOrCreateObjectStore(Diags);` which after such change would always create an on-disk CAS even if it wasn't used (instead of just using an in-memory one). I don't like that it would be very easy to introduce code that inadvertently opened an on-disk CAS every time clang is invoked.
* For round-tripping of options it becomes ambiguous what an empty `CASPath` means, is the option missing or is the option set to an empty string? I tried with changing `CASPath` to optional string but the `MarshallingInfo` mechanism does not seem able to handle optional fields.

Because of this I kept the in-memory/"auto" functionality for now, we could revisit later on. But I did make other simplifications, see https://github.com/apple/llvm-project/pull/6312/commits/dfc2901bf4ed19e49b2458e976fd3f323312c2cf

cachemeifyoucan commented on llvm/include/llvm/CAS/ObjectStore.h:
```diff
@@ -376,6 +376,13 @@ using ObjectStoreCreateFuncTy =
 
+class ActionCache;
+
+/// Create on-disk \p ObjectStore and \p ActionCache instances based on \p
```
Why don't just put this in UnifiedOnDiskCache.h?

akyrtzi replied:
I want to keep `UnifiedOnDiskCache` (header and C++ file) free of dependencies to `ObjectStore` or `ActionCache`.
I intend to create a Swift package with a Swift wrapper interface over `UnifiedOnDiskCache`; it will get copies of llvm sources but I'd only need a subset of the files from `lib/CAS`.

cachemeifyoucan replied:
I think we either let this function return `UnifiedOnDiskCache` and switch away from using ObjectStore and ActionCache in clang, or maybe just rename this function for now to something that doesn't get associated with `UnifiedOnDiskCache`

akyrtzi replied:
> I think we either let this function return UnifiedOnDiskCache and switch away from using ObjectStore and ActionCache in clang

clang needs to be using ObjectStore/ActionCache APIs since there's going to be another implementation of these for clang to use, via the plugin.

> or maybe just rename this function for now to something that doesn't get associated with UnifiedOnDiskCache

My concern with that is that it will become unclear what the difference is between this function and `createOnDiskCAS`.

I'll pull `createOnDiskUnifiedCASDatabases` out of `ObjectStore.h` and put it in a separate header.

akyrtzi replied:
I moved `createOnDiskUnifiedCASDatabases` out of `ObjectStore` here: https://github.com/apple/llvm-project/pull/6312/commits/476b6444343d7c01ace21bd92aa5c723dd3864d7

cachemeifyoucan reviewed (approved):
LGTM

akyrtzi committed:
[clang/CAS] Simplify `clang::CASOptions`

Replace `getOrCreateObjectStore` and `getOrCreateActionCache` with one function, `getOrCreateDatabases`.
Also re-enable the `CASOptionsTest` unit tests, by using a new function `isOnDiskCASEnabled` to check for on-disk
feature, instead of using the macro (which was not defined for that target).

akyrtzi committed:
[llvm/CAS] Move `createOnDiskUnifiedCASDatabases` function to a separate header than `ObjectStore.h`

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan commented:
LGTM

akyrtzi commented:
@swift-ci Please test
